sycl/include/vecmem/memory/sycl/details/memory_resource_base.hpp

```diff
@@ -0,0 +1,41 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// Local include(s).
+#include "vecmem/memory/resources/memory_resource.hpp"
+#include "vecmem/utils/sycl/queue_wrapper.hpp"
+
+namespace vecmem::sycl::details {
+
+   /// SYCL memory resource base class
+   ///
+   /// This class is used as base by all of the oneAPI/SYCL memory resource
+   /// classes. It holds functionality that those classes all need.
+   ///
+   class memory_resource_base : public memory_resource {
+
+   public:
+      /// Constructor on top of a user-provided queue
+      memory_resource_base( const queue_wrapper& queue = { "" } );
```

Comment:
I not sure I really like this, but I understand it's necessary due to the way SYCL work, and I don't see another way of doing. How do we envision this working in actual SYCL code? Does the user intersperse their standard SYCL queue operations with creating `vecmem` vectors and such? For example:

```cpp
sc::sycl::queue q;
vecmem::sycl::device_memory_resource mem(q);
vemem::vector vec(&mem);
vec = {41, 51, 79};
q.submit([&](...){...});
```

Is that roughly the design we're going for?

Reply:
In the diffs it's pretty difficult to see this, but this is one of the things that I liked about this design the most. The SYCL test does the following in this PR:

```c++
   // Create the SYCL queue that we'll be using in the test.
   cl::sycl::queue queue{ vecmem::sycl::device_selector() };
   std::cout << "Using device:" << std::endl;
   std::cout << "    name: "
             << queue.get_device().get_info< cl::sycl::info::device::name >()
             << std::endl;
   std::cout << "  vendor: "
             << queue.get_device().get_info< cl::sycl::info::device::vendor >()
             << std::endl;
   std::cout << " version: "
             << queue.get_device().get_info< cl::sycl::info::device::version >()
             << std::endl;

   // Set up the memory resource for the test.
   vecmem::sycl::shared_memory_resource resource( &queue );

   // Create an input and an output vector.
   auto inputvec = make_input_vector( resource );
   auto outputvec = make_output_vector( resource );
```

https://github.com/acts-project/vecmem/blob/SyclMemoryResource-main-20210303/tests/sycl/test_sycl_containers.sycl#L36-L54

I was trying to achieve exactly what you wrote, to be able to seamlessly give a user-created `cl::sycl::queue` object to the memory resources, while also allowing the resources to be created without any SYCL code dependence as well, in case that is necessary.